src/expr/impl/src/aggregate/jsonb_agg.rs

```diff
@@ -13,72 +13,39 @@
-#[aggregate("jsonb_agg(*int) -> jsonb")]
-#[aggregate("jsonb_agg(*float) -> jsonb")]
-#[aggregate("jsonb_agg(decimal) -> jsonb")]
-#[aggregate("jsonb_agg(serial) -> jsonb")]
-#[aggregate("jsonb_agg(int256) -> jsonb")]
-#[aggregate("jsonb_agg(date) -> jsonb")]
-#[aggregate("jsonb_agg(time) -> jsonb")]
-#[aggregate("jsonb_agg(timestamp) -> jsonb")]
-#[aggregate("jsonb_agg(timestamptz) -> jsonb")]
-#[aggregate("jsonb_agg(interval) -> jsonb")]
-#[aggregate("jsonb_agg(varchar) -> jsonb")]
-#[aggregate("jsonb_agg(bytea) -> jsonb")]
-#[aggregate("jsonb_agg(jsonb) -> jsonb")]
-fn jsonb_agg(state: &mut JsonbArrayState, input: Option<impl ToJsonb>) -> Result<()> {
-    // FIXME(runji):
-    // None of the input types we currently support depend on `data_type` in `add_to`.
-    // So we just use a dummy type here.
-    // To get the correct type, we need to support `ctx: &Context` argument in `#[aggregate]`.
-    let data_type = &DataType::Int32;
-
-    input.add_to(data_type, &mut state.0)?;
+#[aggregate("jsonb_agg(*) -> jsonb")]
+fn jsonb_agg(
+    state: &mut JsonbArrayState,
+    input: Option<impl ToJsonb>,
+    ctx: &Context,
+) -> Result<()> {
+    input.add_to(&ctx.arg_types[0], &mut state.0)?;
     Ok(())
 }
 
 /// Collects all the key/value pairs into a JSON object.
-/// // TODO: support "any" type key
 /// // Key arguments are coerced to text;
 /// value arguments are converted as per `to_jsonb`.
 /// Values can be null, but keys cannot.
-#[aggregate("jsonb_object_agg(varchar, boolean) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, *int) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, *float) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, decimal) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, serial) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, int256) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, date) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, time) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, timestamp) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, timestamptz) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, interval) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, varchar) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, bytea) -> jsonb")]
-#[aggregate("jsonb_object_agg(varchar, jsonb) -> jsonb")]
+#[aggregate("jsonb_object_agg(varchar, *) -> jsonb")]
 fn jsonb_object_agg(
     state: &mut JsonbObjectState,
     key: Option<&str>,
     value: Option<impl ToJsonb>,
+    ctx: &Context,
 ) -> Result<()> {
-    // FIXME(runji):
-    // None of the input types we currently support depend on `data_type` in `add_to`.
-    // So we just use a dummy type here.
-    // To get the correct type, we need to support `ctx: &Context` argument in `#[aggregate]`.
-    let data_type = &DataType::Int32;
-
     let key = key.ok_or(ExprError::FieldNameNull)?;
     state.0.add_string(key);
-    value.add_to(data_type, &mut state.0)?;
+    value.add_to(&ctx.arg_types[1], &mut state.0)?;
```

Comment:
Off-topic: this API looks really weird😇

Reply:
😇For perfomance reasons, the jsonb library uses a `builder` struct. Let's keep it unchanged for now.